lib/charms/alertmanager_k8s/v0/consumer.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+""" # AlertmanagerConsumer library
+
+This library is design to be used by a charm consuming the alertmanager-k8s relation.
+"""
+
+
+import ops
+from ops.framework import StoredState, EventBase, EventSource
+from ops.relation import ConsumerBase
+
+from typing import List
+import logging
+
+LIBID = "abcdef1234"  # Unique ID that refers to the library forever
+LIBAPI = 0    # Must match the major version in the import path.
+LIBPATCH = 1  # The current patch version. Must be updated when changing.
+
+logger = logging.getLogger(__name__)
+
+
+class AlertmanagerConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Alertmanager.
+    This consumer auto-registers relation events on behalf of the user and communicates information
+    directly via `_stored` TODO: have a documented contract and act on it in the "available" hook.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+    _stored: StoredState
+
+    def __init__(self, charm: ops.charm.CharmBase, relation_name, consumes, multi=False):
+        super().__init__(charm, relation_name, consumes, multi)
+        self.charm = charm
+        self._consumer_relation_name = relation_name  # from consumer's metadata.yaml
+        # self._provider_relation_name = "alerting"  # from alertmanager's metadata.yaml
+
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_changed,
+                               self._on_relation_changed)
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_departed,
+                               self._on_relation_departed)
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_broken,
+                               self._on_relation_broken)
+
+        self._stored.set_default(alertmanagers={})
+
+    def _on_relation_changed(self, event: ops.charm.RelationChangedEvent):
```

Comment:
I am curious to understand why you choose to make the consumer (rather than the provider) responsible for aggregating "public address" of all Alertmanager peers. In the same vein I would be interested in your rationale for using stored state in the charm library to store this list rather than application data.

Reply:
This reflects my current understanding, and was tested successfully on 2.9.3 and 2.9.5.
I opened an [issue](https://github.com/canonical/alertmanager-operator/issues/8) to refactor this into relation data.